roadmap updates

ngoldman opened the pull request:
Starting a conversation about API changes here. This means breaking changes.

Trying to keep this focused on immediate needs and API improvements and limit scope creep as much as possible.

A lot of ideas come from the discussion in https://github.com/koopjs/koop/issues/195.

@koopjs/dev feedback welcome.

ngoldman commented:
I'm pulling out the points from July 2015 and placing them here.

```md
## Summer 2015

### Changes
* FeatureService optimizations
  * Make smarter requests to the DB based on queries
  * clean up the logic to be easier to grok
* Refactor the lib/* files on a whole
  - We want pull out as much non-koop specific code as possible into separate modules
* Add support for the simple style spec - https://github.com/mapbox/simplestyle-spec
* Address consistency issues in method calls to the Cache from providers
  - be clear and consistent about passing table names instead of "types" and "keys"
```

We need to clarify and decide what is still relevant above, what is feasible in the short-term, what belongs in a v3, etc.

ngoldman commented:
Removed "A koop desktop app built with something like Electron" as it is a nice-to-have but pretty clearly out of scope. We've got https://github.com/koopjs/koop-ui as a starting point for that discussion, so I think it's best to keep that out of the main Koop roadmap for now.

ngoldman commented:
New document can be previewed more easily here:

https://github.com/koopjs/koop/blob/v3-roadmap/ROADMAP.md

ajturner commented:
@ngoldman superb work on the roadmap outline. This is a great approach to take given the complexity, maturity and community interest in where Koop should go.

Perhaps too big for v3 - but it would be interesting to discuss the _lingua franca_ within Koop. Is GeoJSON the optimal choice or is there a more efficient (binary?) option? If GeoJSON is used, are Koop caches taking advantage of indexes for space/time/attributes?

chelm commented:
Looks awesome @ngoldman, some questions:

* Is there any issue with having multiple "exports" installed that may have conflicting query params? 
* How will we deal with "sources" that need to have query params directly passed back to the source API?
  * would we expect a source's extract method to be passed the query string?
  * how does that interact with the installed export plugins?   

@ajturner Good question about formats. I like GeoJSON since its standard, its json, and can be handed to caches w/o worrying about how it'll be stored, indexed, and queries (as opposed to binary options that may be supported in one cache but not another). Its a good place to start and a cache could make the call to store data in a more efficient structure. In terms of the caches taking advantage of indexing: the PostGIS cache indexes things very well (though for things like live tile creation its way too slow), and the ES cache is not fully indexing the geometries but could.

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+
+Provide better support for `plugins`. Right now the model for creating a plugin is pretty poor, as illustrated by https://github.com/koopjs/koop/issues/195.
+
+Move away from `provider` in favor of three new types of plugins: `source`, `export`, and `interface`.
```
I think this is too ambitious for a v.next

Instead, lets focus on plugins and allowing providers to inherit things from Koop core or a plugin without code to enable it.

E.g. write no provider code and get: exports, geohash, feature service, SODA querying etc, "for free"

ngoldman replied:
It is a bit far-reaching, but I think it's important to have a long-term plan for what an ideal Koop looks like rather than letting development be guided by feature creep.

With that said, I agree we should focus in the short-term on your above recommendations. Perhaps we can use that as a template for a v3 and call this v4 or vN.

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+
+* Support for TopoJSON ([#199](https://github.com/koopjs/koop/issues/199))
+* File Geodatabase support
+* Removing ogr2ogr as a dependency for exporting file formats
```
why remove OGR?

ngoldman replied:
This was previously in the roadmap, lumped it under exports as it's part of that problem area. Don't remember the why at the moment.

ngoldman replied:
@chelm or @ajturner (or anyone in @koopjs/dev), do you remember the reasoning behind adding removal of OGR as a goal to the roadmap? I would imagine the idea is to reduce dependencies and do as much work as possible in javascript. Was there another reason?

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+* fetch (primary method used by koop core to request data from source)
+
+### Export
+
```
Refactoring the export code is a beast in and of itself. Needs to be done, but keep in mind it's going to be significant work. I agree that we should extract that code.

ngoldman replied:
Agreed. Maybe this can be a more short-term focus as it does seem to be a necessity. This could also give us independent modules for transforming GeoJSON into a Shapefile, CSV, KML, FGDB etc (note that some of these already exist) that would be useful outside of the project.

dmfenton replied:
That would be really nice. It would also be nice to be able to plugin an export component that uses Spark, for example

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+
+### Interface
+
+Defines methods for querying GeoJSON via a third-party API. This means Feature Service, Soda2, OData, etc. This requires a lot of refactoring but would yield a lot of benefits and has the potential to set a standard interface for querying GeoJSON.
```
This would be huge. Let's start with making koop's feature service querying and output into a plugin. Then set up Koop such that every provider get's to access to this code if the plugin is enabled. That would be my absolute top priority for any API changes.

ngoldman replied:
I agree it's a very big change. The idea is mainly in response to trying to find a way to serve the use case @timwis brought up in [#195](https://github.com/koopjs/koop/issues/195) (How to build other API output formats?). This does raise the question of whether we should aim to support other API output formats. Is this something we want Koop to do?

I think starting with modularizing Koop's feature service querying and output is an excellent idea.

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+
+e.g.
+
+koop-export-geojson
```
I'm not sure we need all of these modules for exporting. There's so much shared code in the process, sometimes you are just changing a single parameter in OGR

ngoldman replied:
Making a module to encapsulate logic for a specific use case is what makes node really useful. I don't think there's a disadvantage to keeping modules small and focused, even if does mean changing a single parameter. If a lot of code needs to be shared we could create a base `koop-export` class for the export modules to inherit from.

dmfenton commented on ROADMAP.md:
````diff
@@ -2,23 +2,100 @@
+```
+
+Methods:
+
````
what about checking for expiration?
how does this support pass-through transformation.

E.g. I make a request like koop/googlePlaces/FeatureService/0/query?where=1=1&text=pizza
What I want is for the request to pass through the google places source, and get transformed into a feature service or geojson

dmfenton replied:
In general, I think you need to be a lot more specific about what the API endpoints actually look like for different provider use cases.

ngoldman replied:
re: expiration: How are we checking for expiration right now?

re: pass-through transformation: Is this something we're doing right now? I think this is similar to @chelm's question (section 3 above). Is this a reasonable thing for Koop to support in addition to everything else it's doing? In-memory transformation can be very costly and I'm not sure it's part of the scope of what Koop's supposed to do.

ngoldman replied:
Re: more specificity about API endpoints for different provider use cases.

Sure. I think that can be fleshed out as the roadmap progresses. I would love your help on that one.

dmfenton commented on ROADMAP.md:
```diff
@@ -2,23 +2,100 @@
+
+This can be thought of more or less like so:
+
+`server/koop/:source{/**}/:item{.:export,/:interface}`
```
How do providers translate from the interfaces to their underlying APIs?

ngoldman replied:
In the design I outlined, there are no providers, so I'm not sure what this means.

dmfenton replied:
In the nomenclature you've set up:
How does a request flow from an interface to a source to the actual underlying API?

dmfenton commented:
Nice work @ngoldman my comments are in line.

The bottom line:
- I agree in principle but I think this is too big
- Let's focus on plugins, error reporting and provider "inheritance"

ngoldman commented:
Thanks all for the feedback. @dmfenton I tried to answer all your comments in line.

## 1. GeoJSON vs alternative *lingua franca* (binary)
re: @ajturner's question:

> Is GeoJSON the optimal choice or is there a more efficient (binary?) option? If GeoJSON is used, are Koop caches taking advantage of indexes for space/time/attributes?

GeoJSON has been the defacto format for representing spatial data in Koop since its inception. Since it is the most common and easiest-to-use spatial format for the web, and Koop is meant for the web, I think the decision makes sense. In terms of performance, a binary format would probably be much faster. That would also require a full rewrite of Koop, or at the very least a very significant long-term refactor. I think there is merit in making GeoJSON more performant by investigating using streams for parsing and transforms to keep things out of memory as much as possible, and looking into the [`jsonb` postgres datatype](http://www.postgresql.org/docs/9.4/static/datatype-json.html) for indexing. This helps us, is valid R&D, and moves geospatial web technology forward.

## 2. Conflicting query params for exports
re: @chelm's question:

> Is there any issue with having multiple "exports" installed that may have conflicting query params?

I'm not sure exactly what you mean -- how can exporting something to a static format conflict with exporting something else? I'm imagining the export function would act on cached data passed to it via koop.

## 3. Passing query params directly to source
re: @chelm's question:

>* How will we deal with "sources" that need to have query params directly passed back to the source API?
>  * would we expect a source's extract method to be passed the query string?
>  * how does that interact with the installed export plugins?

In this design the source module and the export module would know nothing about each other.  The idea would be to completely decouple extraction of data from delivery of data. If it's necessary for us to pass query params directly to the source API, this design would not facilitate that. Is there a such a scenario we need to support? That would not be very performant (extracting, transforming, and exporting data in-memory), and if the transformation isn't necessary then Koop isn't really needed in that scenario.

I was imagining the source's extract method would attempt to extract *all* data from a given endpoint, cache it, then the exports and interfaces could do whatever they want with it.

ngoldman commented:
I think @dmfenton is right that we need to focus more on basic plugin support, error reporting, and provider inheritance in the short-term. Those don't necessarily need a major version bump. The idea with this design document was to address the all needs I've seen so far as best as possible. It doesn't have to be v3. It may be entirely wrong. But I think there's value in planning beyond the most immediate problems.

I can add the more immediate needs to the roadmap and please feel free to add commits and edit as you see fit.

dmfenton commented:
> Is there a such a scenario we need to support?
Absolutely, take for example koop-yelp and a potential koop-google-places. We'll never be able to extract all the data we need from these very performant APIs. It also may be against terms to cache data.

> That would not be very performant (extracting, transforming, and exporting data in-memory), and if the transformation isn't necessary then Koop isn't really needed in that scenario.
Transformation is still necessary for examples like the two I just listed. How else do you get restaurants from Zillow into a featureservice?

dmfenton commented:
>  It doesn't have to be v3. It may be entirely wrong. But I think there's value in planning beyond the most immediate problems.

The long term roadmap is very important and this is a good forum for it. Fortunately, the short term needs are very clear. We can continue to discuss longer-term work while we go forward.

timwis commented:
This is awesome, guys! I'm glad to see discussion around interfaces. After writing [soda-postgres](https://github.com/timwis/soda-postgres) to work against a Postgres database I have a better idea of the kind of query translations involved, and I think only basic filtering would be possible (or perhaps reasonable is the right word) without the interface refactoring discussed above. 

Are you guys going to be at the CfA conference or transparency camp? I'll be there and would love to chat about this. I feel like I could convey so much more verbally or on a whiteboard :P

ngoldman commented:
@timwis I won't be at CfA summit but I might be able to make it to JS.Geo in September (http://www.jsgeo.com/). I believe it will be in your neck of the woods this year :)

jgravois commented:
@timwis i'm definitely not the *most* knowledgeable, but i'll definitely be at CforA Summit and would love to chat.

timwis commented:
I'll be at JS.Geo too @ngoldman - looks like it's in October though. Swore it was Sept last I looked too.

ngoldman commented:
@timwis weird, I thought so too. @chelm did the date for js.geo change? Also is the source for the site on github? There's a link to jsconf 2012 at the bottom for some reason.

chelm commented:
@ngoldman @timwis js.geo has always been October 8th, and is currently sold out, but might have 25 more tix coming soon.

The link is for the code of conduct credit that was pulled from jsconf

dmfenton commented:
@timwis I'll be at transparency camp, would love to meet up.

ngoldman commented:
I've updated this PR to try to cover what's been talked about so far. Also added a mention about our intent to break out workers, which has been talked about before.

---

## v2.x / Minor Changes

To be completed in the near future - September/October 2015

### Improvements

#### Error handling, reporting, logging

* https://github.com/koopjs/koop/issues/209
* https://github.com/koopjs/koop/issues/228

#### Better plugin support

https://github.com/koopjs/koop/issues/195

##### make preview routes a plugin

Identical or near-identical JS & HTML is being duplicated across providers for previews. If we improve plugin support enough in v2, we should be able to remove preview code from all providers and make a simple plugin that adds a preview route to every provider. I think this would be a good candidate for designing a better plugin strategy.

### Changes

#### Provider inheritance

via koop-provider module

* https://github.com/koopjs/koop-provider
* https://github.com/koopjs/koopjs.github.io/issues/10

#### Worker & Export manangement

Exporter & ExportWorker should be broken out into separate modules.

## v3

### Removals

* remove from `lib/`: Exporter, ExportWorker, BaseModel, BaseController, FeatureServices, Query, & SpatialReference

These modules will be deprecated in v2 once the koop-provider, koop-exporter, and koop-worker modules are production ready.

---

Everything from the theoretical v3 has been moved to a  "Theoretical / tentative discussion" section at the bottom.

If I've missed anything or there are any issues with what's in there, please let me know. Otherwise I think this is ready to merge.

dmfenton commented:
I'm happy with this. Anyone else want to weigh in? 
@ajturner @chelm perhaps?

chelm commented:
@dmfenton Im good with this. You all are killling it.

ngoldman commented:
I heard a happy and a good. Merging
